src/PIL/PpmImagePlugin.py

```diff
@@ -49,25 +49,39 @@ class PpmImageFile(ImageFile.ImageFile):
     format = "PPM"
     format_description = "Pbmplus image"
 
-    def _token(self, s=b""):
+    def _read_token(self, token=b""):
+        def _ignore_comment():  # ignores rest of the line; stops at CR, LF or EOF
+            while True:
+                c = self.fp.read(1)
+                if c in b"\r\n":
+                    break
+
+        while True:  # read until non-whitespace is found
+            c = self.fp.read(1)
+            if c == b"#":  # found comment, ignore it
+                _ignore_comment()
+                continue
+            if c in B_WHITESPACE:  # found whitespace, ignore it
+                if c == b"":  # reached EOF
+                    raise EOFError("Reached EOF while reading header")
+                continue
+            break
+
+        token += c
+
         while True:  # read until next whitespace
             c = self.fp.read(1)
-            if not c or c in b_whitespace:
+            if c == b"#":
+                _ignore_comment()
+                continue
+            if c in B_WHITESPACE:  # token ended
                 break
-            if c > b"\x79":
-                raise ValueError("Expected ASCII value, found binary")
-            s = s + c
-            if len(s) > 9:
-                raise ValueError("Expected int, got > 9 digits")
-        return s
+            token += c
+        return token
 
     def _open(self):
-
-        # check magic
-        s = self.fp.read(1)
-        if s != b"P":
-            raise SyntaxError("not a PPM file")
```

Comment:
Why remove this block?

Reply:
The [docs](https://pillow.readthedocs.io/en/stable/handbook/writing-your-own-file-decoder.html) state that Pillow:

> ...loops over the available image plugins in the loaded order, and calls the plugin’s `_accept` function with the first 16 bytes of the file. If the `_accept` function returns true, the plugin’s `_open` method is called...

I think the relevant lines are these:

https://github.com/python-pillow/Pillow/blob/2d7c543a2f2d853c8b69425040d36411eb0e2e39/src/PIL/Image.py#L2911-L2912

The `_accept` method already checks the magic numbers. I removed that block because it looked like dead code to me. That assumption might be wrong. My intention was to discuss that and other points before submitting.

I see that other plugins (BMP and JPEG at least) also recheck the magic, either manually (PPM) or calling `_accept`, so I'm probably missing something.

Reply:
`_accept` is only guaranteed to be called if you're going through `Image.open()` and relying on the plugin registration. I can also do this to load an image:

```python
import io
from PIL import PpmImagePlugin
PpmImagePlugin.PpmImageFile(fp=io.BytesIO(b'P5\n2 2\n255\n\x80\x00\x00\xFF')).show()
```

That calls `_open` directly, without calling `_accept` first.

Reply:
So that's what I had missed!

Reply:
The SyntaxError was restored in https://github.com/python-pillow/Pillow/pull/5121/commits/d2ad27d70a00efa98560cf5036326abcae850d5e